#!/bin/bash
#$ -N applybqsr
#$ -ckpt restart
#$ -q som,pub64,free64,asom
#$ -pe make 64
#$ -t 1-178


#############
## Docstring
#############

## before running this script
## move to this directory 
## move eo files from baserecal to a subfolder
## /dfs3/som/dalawson/drb/deepcelllineage/mitolin/data/gen/nguyen_nc_2018/ \
    ## 20191031-bqsr/erroroutput/
## run this script from this ^ directory
## so e & o files get deposited there

## change -t ^ to the samples that you want to run the script on
    ## for testing, use `-t 2-4`
    ## for all, count the number of lines: 
    ## `wc -l dupsmarkedlist.txt'

## to run a script on hpc use `qsub path/to/filename.sh`
    ## `qsub /dfs3/som/dalawson/drb/deepcelllineage/mitolin/src/gatkpp-5-applybqsr-v1.1.0.sh`
## to check on the script's status: `qstat -u dalawson`
## to stop a submitted project by job-ID number: `qdel job-ID-number`


################################################
## load modules, create variables & directories
################################################

## add packages to PATH
module load java/1.8.0.111          # language of gatk & picard-tools
module load gatk/4.1.3.0            # includes picard tools

## create path variables to access data 
path2datadir='/dfs3/som/dalawson/drb/deepcelllineage/mitolin/data/'
path2ref=${path2datadir}'ref/broad/bundles/b38/v0/'
path2gen_nguyen2018=${path2datadir}'gen/nguyen_nc_2018/'
# path to bams generated by gatkpp-3-markdup*.sh
path2dupsmarked=${path2gen_nguyen2018}'20191030-markdup/output/dupsmarked/'

## the path below has list: dupsmarkedlist.txt
path2list=${path2dupsmarked}

## these directories were made by gatkpp-4-baserecal
## these paths are needed to access & deposit data
path2output=${path2gen_nguyen2018}'20191031-bqsr/output/'
path2baserecaltables=${path2output}'baserecaltables/'

## create new paths to deposit data
path2bqsrbams=${path2output}'bqsrbams/'

## make directories for each new 'deposit data' path above
mkdir $path2bqsrbams

## make .keep files so new folders are tracked by git
touch $path2bqsrbams/.keep
touch $path2qchrfil/.keep

## create variables for list of bam files
dupsmarkedlist=${path2list}'dupsmarkedlist.txt'

## create a name variable
## note this name has '.bam' extension included
## e.g. i1-lib001-A01.bam
namewext=`head -n $SGE_TASK_ID $dupsmarkedlist | tail -n 1 | cut -f1`

## remove ext from name
## e.g. name=i1-lib001-A01
name=${namewext%'.bam'}


#############################
## gatk pre-processing con't
#############################


## apply base quality score recalibration to bams
    ## https://software.broadinstitute.org/gatk/documentation/tooldocs/current/org_broadinstitute_hellbender_tools_walkers_bqsr_ApplyBQSR.php
    ## recalibrates the base qualities of the input reads 
    ## based on the recalibration table produced by the BaseRecalibrator tool 
    ## outputs a recalibrated BAM

gatk ApplyBQSR \
   -R $path2ref'Homo_sapiens_assembly38.fasta' \
   -I $path2dupsmarked$name'.bam' \
   --bqsr-recal-file $path2baserecaltables$name'.table' \
   -O $path2bqsrbams$name'.bam'


## Next run gatkpp-6-AnalyzeCovariates-*.sh


